code/__HELPERS/math/ray/ray.dm

```diff
@@ -0,0 +1,212 @@
+
+	if(direction.x == 0)
+		return point.x == origin.x
+
+	if(direction.y == 0)
+		return point.y != origin.y
+
+	var/c_x = (point.x - origin.x) / direction.x
+	var/c_y = (point.y - origin.y) / direction.y
+	return (c_x == c_y && (!max_distance || c_x <= max_distance ))
+
+//returns rebound angle of hit atom
+//assumes atom is 1x1 octogonal box
+/ray/proc/getReboundOnAtom(var/rayCastHit/hit)
+	//calc where we hit the atom
+	var/vector/hit_point = hit.point_raw
+	var/vector/hit_atom_loc = atom2vector(hit.hit_atom) + new /vector(0.5, 0.5)
+
+	var/vector/hit_vector = hit_point - hit_atom_loc
+
+	//we assume every atom is a octogonal, hence we use all_vectors
+	var/smallest_dist = 2 //since all vectors are normalized, the biggest possible distance is 2
+	var/vector/entry_dir = null
+
+	for(var/vector/dir in all_vectors)
+		var/vector/delta = dir.chebyshev_normalized() - hit_vector.chebyshev_normalized()
+		var/dist = delta.chebyshev_norm()
+		if(dist < smallest_dist)
+			smallest_dist = dist
+			entry_dir = dir
+
+	return src.direction.mirrorWithNormal(entry_dir)
+
+
+//gets a point along the ray
+/ray/proc/getPoint(var/distance)
+	var/vector/path = direction * distance
+	return origin + path
+
+//inherit and override this for costum logic
+//=> for possible return values check defines at top of file
+/ray/proc/raycast_hit_check(var/atom/movable/A)
+	return RAY_CAST_HIT_CONTINUE
+
+//returns list of raycasthits
+/ray/proc/cast(var/max_distance = RAY_CAST_DEFAULT_MAX_DISTANCE, var/max_hits = RAY_CAST_UNLIMITED_HITS, var/ignore_origin = TRUE)
+	set background = 1 // infinite loop protection gets triggered
+	//calculating a step and its distance to use in the loop
+	var/vector/a_step = direction * RAY_CAST_STEP
+	var/step_distance = a_step.chebyshev_norm()
+
+	//setting up our pointer and distance to track where we are
+	var/vector/pointer = new /vector(0,0)
+	var/distance = 0
+
+	//positions list to easier check if we already found this position (since we are moving in tiny steps, not full numbers)
+	var/list/vector/positions = list()
+
+	//our result
+	var/list/rayCastHit/hits = list()
+
+	while(distance < max_distance)
+		//moving one step further
+		pointer += a_step
+		distance += step_distance
+
+		//calculating our current position in world space (its two lines cause byond)
+		var/vector/new_position_unfloored = origin + pointer
+		var/vector/new_position = new_position_unfloored.floored()
+
+		//check if we already checked this (floored) vector
+		var/exists = FALSE
+		for(var/vector/V in positions)
+			if(V.equals(new_position))
+				exists = TRUE
+		if(exists)
+			continue
+
+		//check if this is origin and if we should ignore it
+		if(ignore_origin && new_position.equals(origin_floored))
+			continue
+
+		//getting the turf at our current (floored) vector
+		var/turf/T = locate(new_position.x, new_position.y, z)
+
+		//trying hit at turf
+		var/hit_check = raycast_hit_check(T)
+		if(hit_check < RAY_CAST_NO_HIT_CONTINUE) //no_hit_exit
```

Comment:
Assuming `hit_check` is an integer, you can change this with a switch.

Reply:
due to enabling costum hit codes `hit_check` needs to be floating point. check the `/ray/beam_ray` in the top of the `beams.dm` file for an example